### PR TITLE
Moving group membership to new framework

### PR DESCRIPTION
**Pull request by tsorya**
Moving group member ship to new framework

### CODE REVIEW

File: provider/groupmembership/group_membership_resource.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+	// Get the plan:
+	state := &GroupMembershipState{}
+	diags := req.Plan.Get(ctx, state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Wait till the cluster is ready:
+	resource := g.collection.Cluster(state.Cluster.String())
```
Comment (JohnStrunk) [inline provider/groupmembership/group_membership_resource.go:101]: When accessing a TF variable, you should use the Value<Type>() method (i.e., `state.Cluster.ValueString()`)
Other than that, this PR lgtm.

File: provider/groupmembership/group_membership_resource.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+				Description: "user name.",
+				Required:    true,
+			},
+		},
+	}
+	return
+}
+
+func (g *GroupMembershipResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:73]: remove blank line

File: provider/groupmembership/group_membership_resource.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Wait till the cluster is ready:
+	resource := g.collection.Cluster(state.Cluster.String())
+	pollCtx, cancel := context.WithTimeout(ctx, 1*time.Hour)
+	defer cancel()
+	_, err := resource.Poll().
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:104]: I think this polling can be moved to common

File: provider/groupmembership/group_membership_resource.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+		resp.Diagnostics.AddError(
+			"Can't build group membership",
+			fmt.Sprintf(
+				"Can't build group membership for cluster '%s' and group '%s': %v",
+				state.Cluster.String(), state.Group.String(), err,
+			),
+		)
+		return
+	}
+	collection := resource.Groups().Group(state.Group.String()).Users()
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:135]:
```suggestion
	collection := resource.Groups().Group(state.Group.ValueString()).Users()
```

File: provider/groupmembership/group_membership_resource.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+	// Create the membership:
+	builder := cmv1.NewUser()
+	builder.ID(state.User.String())
+	object, err := builder.Build()
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Can't build group membership",
+			fmt.Sprintf(
+				"Can't build group membership for cluster '%s' and group '%s': %v",
+				state.Cluster.String(), state.Group.String(), err,
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:130]:
```suggestion
				state.Cluster.ValueString(), state.Group.ValueString(), err,
```

File: provider/groupmembership/group_membership_resource.go [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+		Predicate(func(get *cmv1.ClusterGetResponse) bool {
+			return get.Body().State() == cmv1.ClusterStateReady
+		}).
+		StartContext(pollCtx)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Can't poll cluster state",
+			fmt.Sprintf(
+				"Can't poll state of cluster with identifier '%s': %v",
+				state.Cluster.String(), err,
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:115]:
```suggestion
				state.Cluster.ValueString(), err,
```

File: provider/groupmembership/group_membership_resource.go [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+			),
+		)
+		return
+	}
+
+	// Create the membership:
+	builder := cmv1.NewUser()
+	builder.ID(state.User.String())
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:123]:
```suggestion
	builder.ID(state.User.ValueString())
```

File: provider/groupmembership/group_membership_resource.go [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+		return
+	}
+	collection := resource.Groups().Group(state.Group.String()).Users()
+	add, err := collection.Add().Body(object).SendContext(ctx)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Can't create group membership",
+			fmt.Sprintf(
+				"Can't create group membership for cluster '%s' and group '%s': %v",
+				state.Cluster.String(), state.Group.String(), err,
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:142]:
```suggestion
				state.Cluster.ValueString(), state.Group.ValueString(), err,
```

File: provider/groupmembership/group_membership_resource.go [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+	// Get the current state:
+	state := &GroupMembershipState{}
+	diags := req.State.Get(ctx, state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Find the group membership:
+	obj := g.collection.Cluster(state.Cluster.String()).Groups().Group(state.Group.String()).
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:165]:
```suggestion
	obj := g.collection.Cluster(state.Cluster.ValueString()).Groups().Group(state.Group.ValueString()).
```

File: provider/groupmembership/group_membership_resource.go [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+		Users().
+		User(state.ID.String())
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:167]:
```suggestion
		User(state.ID.ValueString())
```

File: provider/groupmembership/group_membership_resource.go [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+	get, err := obj.Get().SendContext(ctx)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Can't find group membership",
+			fmt.Sprintf(
+				"Can't find user group membership identifier '%s' for "+
+					"cluster '%s' and group '%s': %v",
+				state.ID.String(), state.Cluster.String(), state.Group.String(), err,
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:175]:
```suggestion
				state.ID.ValueString(), state.Cluster.ValueString(), state.Group.ValueString(), err,
```

File: provider/groupmembership/group_membership_resource.go [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+	// Get the state:
+	state := &GroupMembershipState{}
+	diags := req.State.Get(ctx, state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Send the request to delete group membership:
+	obj := g.collection.Cluster(state.Cluster.String()).Groups().Group(state.Group.String()).
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:204]:
```suggestion
	obj := g.collection.Cluster(state.Cluster.ValueString()).Groups().Group(state.Group.ValueString()).
```

File: provider/groupmembership/group_membership_resource.go [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+		Users().
+		User(state.ID.String())
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:206]:
```suggestion
		User(state.ID.ValueString())
```

File: provider/groupmembership/group_membership_resource.go [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+	_, err := obj.Delete().SendContext(ctx)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Can't delete group membership",
+			fmt.Sprintf(
+				"Can't delete group membership with identifier '%s' for "+
+					"cluster '%s' and group '%s': %v",
+				state.ID.String(), state.Cluster.String(), state.Group.String(), err,
```
Comment (bardielle) [inline provider/groupmembership/group_membership_resource.go:214]:
```suggestion
				state.ID.String(), state.Cluster.ValueString(), state.Group.ValueString(), err,
```

**Pull request merged**